_posts/2016-09-29-Rust-1.12.md

```diff
@@ -0,0 +1,324 @@
+
+Initially, this error design was built to aid borrow-checking errors, but we
+found, like the error above, the format can be broadly applied to a wide variety
+of errors. If you would like to learn more about the design, check out the
+[previous blog post on the subject][err].
+
+[err]: https://blog.rust-lang.org/2016/08/10/Shape-of-errors-to-come.html
+
+## MIR code generation
+
+The new Rust "mid-level IR", usually called "MIR", gives the compiler a simpler
+way to think about Rust code than it's previous way of operating entirely on the
+Rust abstract syntax tree. It makes analysis and optimizations possible that
+have historically been difficult to implement correctly. The first of many
+upcoming changes to the compiler enabled by MIR is a rewrite of the pass that
+generates LLVM IR, what `rustc` calls "translation", and after many months of
+effort the MIR-based backend has proved itself ready for prime-time.
+
+TODO: Improve code complexity: size of old trans vs size of new trans.
+
+While for the most part this new backend is simply laying the groundwork for
+future improvements, this release does include one long-awaited refinement to
+the run-time representation of Rust types that results in both compile-time wins
+and runtime wins: the removal of "drop flags", AKA "non-zeroing drop", as
+specified in [RFC 320].
+
+To maintain the semantics of Rust's `Drop` implementation, we always need to
+check whether a type must be dropped when it goes out of scope. Historically,
+this has been done by adding a special flag to all types that implement `Drop`,
+but this resulted in significant extra memory being used at runtime.
+
+MIR exposes perfect information about the program's control flow, so the
+compiler knows exactly whether types are moved or not. This means that it knows
+statically whether or not the value's destructor needs to be run. In cases where
+a value may or may not be moved at the end of a scope, the compiler now simply
+uses a single bitflag on the stack, which is in turn easier for optimization
+passes in LLVM to reason about. The end result is less work for the compiler and
+less bloat at runtime.
+
+The improvements in compile times of some workloads due to this change alone are
+significant:
+
+|               | 1.11   | 1.12  | compile time |
+|---------------|--------|-------|--------------|
+| inflate-0.1.0 | 10.17s | 4.32s | 2.35x faster |
+| hyper-0.5.0   | 24.03s | 5.23s | 4.58x faster |
+| hellworld     | 0.17s  | 0.14s | 1.20x faster |
+| regex 0.1.30  | 9.63s  | 2.39s | 4.03x faster |
+
+TODO: Future improvements.
+
+TODO: non-zeroing drop PR https://github.com/rust-lang/rust/pull/33622
+
+Compile times:
+
+|               | 1.11   | 1.12  | % change |
+|---------------|--------|-------|----------|
+| inflate-0.1.0 | 10.17s | 4.32s | -57.5%   |
+| hyper.0.5.0   | 24.03s | 5.23s | -78.2%   |
+| helloworld    | 0.17s  | 0.14s | -16.4%   |
+| regex.0.1.30  | 9.63s  | 2.39s | -75.2%   |
```

Comment:
:+1: I actually didn't totally realize that, i was just transcribing from the irccloud link